File: Java/VelocidadGestores3/src/main/java/database/connections/ConexionBase.java
package database.connections;

import lombok.Getter;
import lombok.Setter;

import javax.swing.*;
import java.sql.Connection;
import java.sql.SQLException;

public abstract class ConexionBase implements ConexionBD {
    @Setter @Getter
    protected String dbUrl;
    @Setter @Getter
    protected String dbUser;
    @Setter @Getter
    protected String dbPassword;

    protected Connection conexionBD;

    // Método abstracto para que las subclases implementen su propio driver
    protected abstract void cargarDriver() throws ClassNotFoundException;

    // Método abstracto para que cada subclase cree su conexión
    protected abstract Connection crearConexion() throws SQLException;

    // Conexión general, reutilizable en todas las subclases
    public void conectar() {
        try {
            getConexion();
            mostrarMensajeConexion();
        } catch (SQLException e) {
            e.printStackTrace();
            JOptionPane.showMessageDialog(null, "Error de conexión: " + e.getMessage(), "Error", JOptionPane.ERROR_MESSAGE);
        }
    }

    // Método abstracto para mostrar un mensaje personalizado en cada subclase
    protected abstract void mostrarMensajeConexion();

    // Actualiza credenciales y reconecta
    public void actualizarCredenciales(String dbUrl, String dbUser, String dbPassword) {
        try {
            desconectar();
            this.dbUrl = dbUrl;
            this.dbUser = dbUser;
            this.dbPassword = dbPassword;
            conectar();
        } catch (SQLException e) {
            System.err.println("Error al actualizar credenciales: " + e.getMessage());
            this.dbUrl = null;
            this.dbUser = null;
            this.dbPassword = null;
        }
    }

    // Desconecta la conexión
    public void desconectar() throws SQLException {
        if (conexionBD != null && !conexionBD.isClosed()) {
            conexionBD.close();
            conexionBD = null;
        }
    }

    // Obtiene la conexión, la crea si es necesario
    public Connection getConexion() throws SQLException {
        if (conexionBD != null && !conexionBD.isClosed()) {
            return conexionBD;
        }

        try {
            cargarDriver();
            conexionBD = crearConexion();
            return conexionBD;
        } catch (ClassNotFoundException e) {
            throw new SQLException("Error al cargar el driver de la base de datos.", e);
        }
    }
}

File: Python/VelocidadGestores3/src/database/connections/ConexionBase.py
from abc import abstractmethod
from typing import Optional

from src.database.connections.ConexionBD import ConexionBD


class ConexionBase(ConexionBD):
    _conexion_bd: Optional[object] = None

    def __init__(self, db_host: str, db_port: int, db_name: str, db_user: str, db_password: str):
        self.db_host = db_host
        self.db_port = db_port
        self.db_name = db_name
        self.db_user = db_user
        self.db_password = db_password

    def actualizar_credenciales(self, db_url: str, db_port: int, db_name: str, db_user: str, db_password: str):
        """Actualiza los parámetros de conexión y reconecta la instancia activa."""
        self.desconectar()  # Cerramos la conexión anterior si existe
        self.db_host = db_url
        self.db_port = db_port
        self.db_name = db_name
        self.db_user = db_user
        self.db_password = db_password
        self.conectar()  # Reestablecemos la conexión con las nuevas credenciales

    @abstractmethod
    def crear_conexion(self):
        """Cada subclase debe implementar cómo crear la conexión."""
        pass

    @abstractmethod
    def is_connected(self):
        pass

    @abstractmethod
    def show_connec(self):
        pass

    def conectar(self):
        """Establece la conexión a la base de datos y muestra el mensaje."""
        try:
            conexion = self.get_conexion()
            self.show_connec()  # Muestra mensaje gráfico si la conexión fue exitosa
            return conexion
        except Exception as e:
            print(e)

    def desconectar(self):
        """Cerrar la conexión si está abierta."""
        if self._conexion_bd:
            try:
                self._conexion_bd.close()
            except Exception as e:
                print(f"Error al cerrar la conexión: {e}")
            finally:
                self._conexion_bd = None

    def get_conexion(self):
        """Retorna la conexión actual o la crea si es necesario."""
        if self._conexion_bd and self.is_connected():
            return self._conexion_bd  # Retorna la conexión actual si está activa

        try:
            self._conexion_bd = self.crear_conexion()  # Crea una nueva conexión
            return self._conexion_bd
        except Exception as e:
            raise RuntimeError(f"Error al establecer conexión: {e}")

    def manejar_error(self, mensaje: str):
        """Manejo de errores común a todas las clases."""
        print(mensaje)  # Puedes cambiarlo por logging
        raise RuntimeError(mensaje)
